### ISSUE TITLE
Enabling SemanticDB skips WartRemover on Scala 2.12.17

### ISSUE THREAD

**Issue by bmarker** [scalameta/scalameta#3099]
Unfortunately, I do not have a simple sharable reproduction case.

Scala: 2.12.17
sbt: 1.8.2

Plugin
```
addSbtPlugin("ch.epfl.scala" % "sbt-scalafix" % "0.10.4")
addSbtPlugin("org.wartremover" % "sbt-wartremover" % "3.0.11")
```

If I enable SematicDB in by `build.sbt` file:
```
ThisBuild / semanticdbEnabled := true
```

Then WartRemover no longer runs on compile. If I have code that compiles except for WartRemover errors, it succeeds instead of failing as expected.

On Scala 2.12.10 (and sbt 1.8.2), I was running an older version of Scalafix with WartRemover OK even with SemanticDB enabled and `scalafixOnCompile`.

I was using:
```
addSbtPlugin("ch.epfl.scala" % "sbt-scalafix" % "0.9.34")
addSbtPlugin("org.wartremover" % "sbt-wartremover" % "2.4.12")
```

And I was running Scalafix on compile with:
```
ThisBuild / semanticdbEnabled := true
ThisBuild / semanticdbVersion := scalafixSemanticdb.revision
ThisBuild / scalafixOnCompile := true
```

I was also able to test my current project with WartRemover 2.4.21 (the only 2.4.x version published for Scala 2.12.17) and the behavior is the same.

Possibly interestingly, `sbt doc` still calls WartRemover.

**Comment by bmarker** [comment]
I had been debating filing this with WartRemover or Scalameta and decided on Scalameta because the Scalafix doc specifically mentions running Scalafix and WartRemover together.

It just occurred to me to check what was defining `semanticdbEnabled`, expecting that it might be Scalafix and not SemanticDB. Seems it's maybe just built into `sbt` (sbt/Keys.class). That may put this squarely on the WartRemover side.

**Comment by bmarker** [comment]
I found out that one of my other plugins was pulling in `silencer-plugin`. I don't know why it was only a problem when `semanticdbEnabled` was true.

**Comment by bmarker** [comment]
After tracking it down to Silencer, I found this:
https://github.com/ghik/silencer/issues/43

Which links to this:
https://github.com/scalameta/metals/issues/1409

**Comment by bmarker** [comment]
Looks to me like the Scala reporters change from 2.13.1 was merged back into 2.12.13 as well but the fix only applies to 2.13.1+.

**Comment by bmarker** [comment]
I believe the fix for issue #1505 should be applied to Scala 2.12.13+ as well, which also has the `FilteringReporter`.

**Comment by bmarker** [comment]
Confirmed it works for Scala 2.12.17. SemanticDB 4.7.6 does not log WartRemover errors but my locally published 4.7.6-SNAPSHOT does.

**Comment by bmarker** [comment]
Should I have access to push a branch? I was trying to create a PR using this git profile from a computer with a different default profile so I'm not sure if it's the multiple profiles or if access is restricted.

If a contributor needs to make the change, all I did was copy the file:
 - https://github.com/scalameta/scalameta/blob/main/semanticdb/scalac/library/src/main/scala-2.13/scala/meta/internal/semanticdb/scalac/SemanticdbReporter.scala

 to the 2.12.13 - 2.12.17 paths like:
- https://github.com/scalameta/scalameta/blob/main/semanticdb/scalac/library/src/main/scala-2.12.17/scala/meta/internal/semanticdb/scalac/SemanticdbReporter.scala

**Comment by tgodzik** [comment]
Sorry for getting back to you so late, could you confirm if https://github.com/scalameta/scalameta/pull/3107 works for you?

**Comment by bradley-marker-ck** [comment]
It doesn't even compile for me. A lot of errors from the `trees*` code.

The first errors are like:
```
[error] /git-repos/tgodzik/scalameta/scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala:837:18: @replacedField can't refer to future versions (current is 4.4.10)
[error]   @replacedField("4.6.0") final def argss: List[List[Term]] = argClauses.map(_.values).toList
```

Then there are some type mismatches and other errors.

I'm running `sbt ++2.12.17 publishLocal` with JDK 8 on macOS.

It works fine on the `scalameta:main` branch.

**Comment by bradley-marker-ck** [comment]
Looking at the Scala 2.12.17 JVM 8 test. The `sbt ++2.12.17 testsJVM/test` command also passes for me on `scalameta/scalameta:main` but fails on `tgodzik/scalameta:apply-fix`.

I see I have `(Oracle Corporation Java 1.8.0_211)` but the CI uses `(Temurin Java 1.8.0_362)`.

**Comment by bradley-marker-ck** [comment]
Same errors using `(Amazon.com Inc. Java 1.8.0_362)`.

**Comment by bradley-marker-ck** [comment]
It looks like the tags on your fork only go to v4.4.10 so the build thinks the current version is 4.4.10 instead of 4.7.6. So the 4.6.0 version looks like a future version instead of a past version (also the 4.6.0 tag would be missing if it needed that).

I'm used to branches instead of forks. Looks like I may need to set it up differently.

**Comment by bradley-marker-ck** [comment]
Not sure it's the right thing to do but adding a local v4.7.6 tag fixed the compile for me.

**Comment by bmarker** [comment]
PR #3107 does not work. I noticed it's not the same reporter change from:
https://github.com/scalameta/scalameta/blob/main/semanticdb/scalac/library/src/main/scala-2.13/scala/meta/internal/semanticdb/scalac/SemanticdbReporter.scala

**Comment by bmarker** [comment]
I'm used to branches instead of forks so I may have done something wrong setting the upstream. When building your fork for PR #1107, the latest tag is v4.4.10. It's not finding the v4.7.6 from scalameta. This results in it using the wrong "current" version value, which broke `@replaceField("4.6.0")`  checks because it thought it was a future version.

I worked around it by adding a `v4.7.6` tag locally.

**Comment by bmarker** [comment]
Now that I know you expected a fork, I can try creating a PR myself. I'm about to go out of town so it may not be for a couple weeks.

**Comment by bmarker** [comment]
I created PR #3108.

**Comment by bmarker** [comment]
My mistake. It seems `sbt` doesn't reliably pick up the latest build when publishing multiple `4.7.6-SNAPSHOT` versions from different paths (the different forks).

After making PR #3107 publish `4.7.7-SNAPSHOT` locally, it appears to work.
PR #3108 works as well (published as `4.7.8-SNAPSHOT` to keep them separate).

**Comment by bmarker** [comment]
There seems to be some aggressive caching or something going on. Even when I publish my different forks as different snapshot versions, it doesn't seem to get the new version when I change the `semanticdbVersion` in my `build.sbt` file.

**Comment by bmarker** [comment]
Never mind. I built my "4.7.8-SNAPSHOT" for my fork off the main branch instead of the PR branch.

Now that I published 4.7.8-SNAPSHOT of the right branch, confirmed PR #3108  works.

**Comment by bmarker** [comment]
After some cache busting, switching to the correct branches, etc., I've confirmed that PR #3107 does not work for me but PR #3108 does.

I thought maybe PR #3107 would have an error count but not log the errors. It doesn't even do that. The compile succeeds (instead of failing with the WartRemover error).

**Comment by bmarker** [comment]
Confirmed that SemanticDB v4.7.7 works for me.

```
// ThisBuild / semanticdbVersion := scalafixSemanticdb.revision
ThisBuild / semanticdbVersion := "4.7.7"
```

I'm not sure what it will take to get the ScalaFix plugin `scalafixSemanticdb` updated.

**Comment by bmarker** [comment]
I see the ScalaFix update merged but it doesn't look to me like it was published yet:
https://github.com/scalacenter/scalafix/commits/main

**Comment by bmarker** [comment]
Looks like ScalaFix 0.11.0 should release with this fix (and Scala 2.12.18 and Scala 2.13.11 support).
